storage/simulation/script.go

```diff
@@ -0,0 +1,310 @@
+	"strings"
+	"text/tabwriter"
+
+	"github.com/cockroachdb/cockroach/util"
+)
+
+// Operation enumerates the possible scripted operations that can occur to a
+// cluster.
+type Operation int
+
+// These are the possible scripted operations.
+const (
+	OpSplitRange Operation = iota
+	OpAddNode
+	OpExit
+	// TODO(brma): add optoinal value to addnode to indicated size
+	// TODO(bram): consider many other operations here.
+	//	OpMergeRangeRandom
+	//	OpMergeRangeFirst
+	//  OpKillNodeRandom
+	//  OpKillNodeFirst
+	//  OpAddStoreRandom
+	//  OpAddStore
+	//	OpKillStoreRandom
+	//	OpKillStoreFirst
+)
+
+var operations = [...]string{
+	OpSplitRange: "splitrange",
+	OpAddNode:    "addnode",
+	OpExit:       "exit",
+}
+
+// String returns the string version of the Operation enumerable.
+func (operation Operation) String() string {
+	return operations[operation]
+}
+
+// operationFromString returns the operation corresponding with the passed in
+// string.
+func operationFromString(s string) (Operation, bool) {
+	for i, op := range operations {
+		if op == s {
+			return Operation(i), true
+		}
+	}
+	return 0, false
+}
+
+// OperationValue enumerates the possible values for a scripted operation.
+type OperationValue int
+
+// These are the possible operations values.
+const (
+	OpValSet OperationValue = iota
+	OpValRandom
+	OpValFirst
+	OpValLast
+)
+
+var operationValues = [...]string{
+	OpValSet:    "",
+	OpValRandom: "random",
+	OpValFirst:  "first",
+	OpValLast:   "last",
+}
+
+// String returns the string version of the OperationValue enumerable.
+func (operationValue OperationValue) String() string {
+	return operationValues[operationValue]
+}
+
+// operationValueFromString returns the operation corresponding with the passed
+// in string. If the value is opValSet, then the value is parsed into the
+// returned int.
+func operationValueFromString(s string) (OperationValue, int, error) {
+	for i, opVal := range operationValues {
+		if opVal == s {
+			return OperationValue(i), 0, nil
+		}
+	}
+	valueNumber, err := strconv.Atoi(s)
+	return OpValSet, int(valueNumber), err
+}
+
+// Action is a single scripted action.
+type Action struct {
+	operation   Operation
+	value       OperationValue
+	valueNumber int
+}
+
+// ActionDetails contains an action and all of the metadata surrounding that
+// action.
+// First is the first epoch in which the action occurs.
+// Every is the interval for how often the action re-occurs.
+// Last is the final epoch in which the action can occur.
+// Repeat is how many times an action occurs at each epoch in which it fires.
+type ActionDetails struct {
+	Action
+	first  int
+	every  int
+	last   int
+	repeat int
+}
+
+func (ad ActionDetails) String() string {
+	var buf bytes.Buffer
+	fmt.Fprintf(&buf, "Operation:%s", operations[ad.Action.operation])
+
+	if ad.Action.value == OpValSet {
+		if ad.Action.valueNumber > 0 {
+			fmt.Fprintf(&buf, "\tValue:%d", ad.Action.valueNumber)
+		} else {
+			fmt.Fprintf(&buf, "\t ")
```

Comment:
no need, with the tabwriter later on.  Also, it would require a flush in here then.